core/src/price_estimation/orderbook_based.rs

```diff
@@ -0,0 +1,149 @@
+use super::price_source::PriceSource;
+use crate::models::{BatchId, TokenId};
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use pricegraph::{Orderbook, TokenPair};
+use primitive_types::U256;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+pub struct PricegraphEstimator {
+    orderbook_reader: Arc<dyn StableXOrderBookReading>,
+}
+
+const ONE_OWL: f64 = 1_000_000_000_000_000_000.0;
+
+impl PriceSource for PricegraphEstimator {
+    fn get_prices<'b>(
+        &'b self,
+        tokens: &'b [TokenId],
+    ) -> BoxFuture<'b, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let batch = BatchId::currently_being_solved(SystemTime::now())?;
+            let (account_state, orders) =
+                self.orderbook_reader.get_auction_data(batch.into()).await?;
+            let orderbook = Orderbook::from_elements(orders.iter().map(|order| {
```

Comment:
Note, that I changed this to no longer reduce the orderbook beforehand, as price estimates should be based on the overlapping orderbook.

Reply:
The first step to estimating prices is to reduce the orderbook. I think from a performance standpoint it makes sense to reduce it first.

Reply:
Can we not estimate prices without reducing the orderbook first? I don't think it makes a huge difference (prices won't swing more than a few % before/after reduction), but technically the solver needs to know the estimated prices to clear the orderbook as is. This should include prices of overlapping trade.

E.g. the price of an esoteric non-liquid token might be infinity if all its supply has been covered by a matching order. However the price that the solver needs to know should be in the overlapping range.

On the other hand, this might anchor the prices for market orders very low (in the extreme case estimating the price at 0 before reducing and at infinity after). This could be bad for the fairness of such orders.

Reply:
I think what Nick means is that the orderbook is reduced internally anyway when you estimate a price with `fill_market_order`. Otherwise we wouldn't be able to calculate a price because the negative cycle makes the cost infinitely negative.
Since the reductions happens anyway it is more performant to do it once before estimating any prices.

Reply:
> Can we not estimate prices without reducing the orderbook first?

The price estimate in its current form, is really about finding an exchange rate such that there is an overlapping order with the current orderbook. When there are negative cycles (i.e. existing overlaps in the orderbook), its unclear which order this overlap will match with, so it gets reduced first.

I agree, however, that for price source estimation, it would be nice to include this as well so that the estimate is closer to the middle of the spread instead of being on one side of it.